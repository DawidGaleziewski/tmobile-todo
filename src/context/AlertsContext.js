import React, { createContext, useState } from "react";

// Hooks
import useAlerts from "../hooks/useAlerts";

// Utils
import { v4 as uuidv4 } from "uuid";

const AlertsContext = createContext(null);

const AlertsProvider = ({ children }) => {
  const [alerts, setAlerts] = useState([]);
  const addAlert = (alertType, alertText) => {
    const id = uuidv4();
    const updatedAlerts = [
      ...alerts,
      {
        text: alertText,
        type: alertType,
        id: id,
      },
    ];
    setAlerts(updatedAlerts);
  };
  const removeAlert = (id) => {
    setAlerts([...alerts].filter((alert) => alert.id !== id));
  };

  useAlerts(alerts, removeAlert);

  return (
    <AlertsContext.Provider
      value={{
        alerts,
        addAlert,
        removeAlert,
        setAlerts,
      }}
    >
      {children}
    </AlertsContext.Provider>
  );
};

export { AlertsContext, AlertsProvider };
