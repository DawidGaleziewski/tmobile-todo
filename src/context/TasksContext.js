import React, { createContext, useState } from "react";

// Hooks
import useAxiosGetCall from "../hooks/useAxiosGetCall";

// Authorization
import { tasksAPIUrl } from "../auth/tasksAPISettings";

const TasksContext = createContext([]);

const TasksProvider = ({ children, setIsloading }) => {
  const [tasksState, setTasksState] = useState([]);
  const [isServerUnreachable, setIsServerUnrechable] = useState(false);

  const updateOneTaskState = (itemID, newItem) => {
    const findItemIndex = (array, itemID) =>
      array.findIndex((item) => item.id === itemID);

    const getUpdatedArray = (array, itemID, newItem) => {
      const arrayCopy = JSON.parse(JSON.stringify(array));
      arrayCopy.splice(findItemIndex(arrayCopy, itemID), 1, newItem);
      return arrayCopy;
    };

    setTasksState(getUpdatedArray(tasksState, itemID, newItem));
  };

  const removeTask = (taskID) => {
    console.log(taskID);
    setTasksState(
      [...tasksState].filter((task) => task.id !== parseInt(taskID))
    );
    console.log(tasksState);
  };

  useAxiosGetCall(
    tasksAPIUrl,
    setIsloading,
    (data, error) => {
      console.log("hooonk", error, data);
      if (error) {
        setIsServerUnrechable(true);
        return setTasksState([]);
      }
      setTasksState(data);
    },
    []
  );

  return (
    <TasksContext.Provider
      value={{
        tasksState,
        setTasksState,
        updateOneTaskState,
        removeTask,
        isServerUnreachable,
      }}
    >
      {children}
    </TasksContext.Provider>
  );
};

export { TasksContext, TasksProvider };
