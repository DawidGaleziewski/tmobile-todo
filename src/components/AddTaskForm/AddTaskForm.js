// Libs
import React, { useState, useContext, Fragment } from "react";
import axios from "axios";
import { css, jsx } from "@emotion/core";

// Components
import InputSlider from "../atoms/inputs/InputSlider/InputSlider";
import FormButton from "../atoms/buttons/FormButton";
import CloseWindowButton from "../atoms/buttons/CloseWindowButton/CloseWindowButton";
import ToggleFormButton from "../atoms/buttons/ToggleFormButton";
import LoadingDefault from "../atoms/loadings/LoadingCircle";
import Checkbox from "../atoms/inputs/Checkbox/Checkbox";

// Authorization
import { tasksAPIUrl, httpHeader } from "../../auth/tasksAPISettings";

// Context
import { TasksContext } from "../../context/TasksContext";
import { AlertsContext } from "../../context/AlertsContext";

// Utils
import setFormChanges from "../../utils/setFormChanges";

// Styles
import { formWrapper, barFormStyle } from "./style/AddTaskFormStyle";

/** @jsx jsx */

const AddTaskForm = () => {
  const [isMobileFormToggled, setIsMobileFormToggled] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const { tasksState, setTasksState } = useContext(TasksContext);
  const { addAlert } = useContext(AlertsContext);

  const initialState = {
    title: "",
    description: "",
    done: false,
  };

  const [formState, setFormState] = useState(initialState);

  const postNewTaskCall = async (formState, tasksState) => {
    setIsLoading(true);
    try {
      const { data } = await axios.post(tasksAPIUrl, formState, httpHeader);
      setIsLoading(false);
      setTasksState([...tasksState, data]);
    } catch (error) {
      setIsLoading(false);
      addAlert(
        "danger",
        "Ups. Unable to add task. Check data and try again",
        3000
      );
      console.log(error);
    }
  };
  const onSubmitHandler = (event) => {
    event.preventDefault();
    postNewTaskCall(formState, tasksState);
  };

  const onChangeHandler = (event) => {
    setFormChanges(event, setFormState, formState);
  };

  const toggleMobileFormHandler = (event) => {
    event.preventDefault();
    setIsMobileFormToggled(!isMobileFormToggled);
  };
  return (
    <div css={formWrapper} data-testid="add-task-form">
      <ToggleFormButton
        buttonText="New task"
        onClickHandler={toggleMobileFormHandler}
      />

      <form css={barFormStyle(isMobileFormToggled)} onSubmit={onSubmitHandler}>
        {isLoading ? (
          <LoadingDefault />
        ) : (
          <Fragment>
            <CloseWindowButton onClickHandler={toggleMobileFormHandler} />
            <Checkbox
              id="taskStatus"
              name="done"
              value={formState.done}
              onChange={onChangeHandler}
            />
            <InputSlider
              id={"taskTitle"}
              type={"text"}
              name={"title"}
              labelText={"task title"}
              value={formState.title}
              onChange={onChangeHandler}
              required={true}
            />
            <InputSlider
              id={"taskDescription"}
              type={"text"}
              name={"description"}
              labelText={"task description"}
              value={formState.description}
              onChange={onChangeHandler}
            />
            <FormButton btnText="Add task" data-testid="submit-btn" />{" "}
          </Fragment>
        )}
      </form>
    </div>
  );
};

export default AddTaskForm;
