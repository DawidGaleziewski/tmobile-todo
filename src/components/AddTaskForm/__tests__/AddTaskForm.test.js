// Libraries
import React, { useContext } from "react";
import ReactDOM from "react-dom";
import { render, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";

// Context
import { TasksProvider } from "../../../context/TasksContext";
import { AlertsProvider } from "../../../context/AlertsContext";

// Components
import AddTaskForm from "../AddTaskForm";

afterEach(cleanup);

it("renders AddTaskForm without crashing", () => {
  const divContainer = document.createElement("div");

  ReactDOM.render(
    <AlertsProvider>
      <AddTaskForm />
    </AlertsProvider>,
    divContainer
  );
  ReactDOM.unmountComponentAtNode(divContainer);
});

it("matches snapshot 1", () => {
  const DOMTree = renderer
    .create(
      <AlertsProvider>
        <AddTaskForm />
      </AlertsProvider>
    )
    .toJSON();

  expect(DOMTree).toMatchSnapshot();
});
