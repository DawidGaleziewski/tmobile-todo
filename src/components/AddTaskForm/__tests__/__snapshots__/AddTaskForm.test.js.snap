// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1 1`] = `
<div
  className="css-25o2qi-AddTaskForm"
  data-testid="add-task-form"
>
  <button
    className="css-1r1xgwo-ToggleFormButton"
    onClick={[Function]}
  >
    New task
  </button>
  <form
    className="css-8enizf-AddTaskForm"
    onSubmit={[Function]}
  >
    <button
      className="close-window-btn css-jh2deu-CloseWindowButton"
      onClick={[Function]}
    >
      <img
        alt="close"
        src="/close-x.png"
      />
    </button>
    <div
      className="input-slider-wrapper"
      data-testid="checkbox"
    >
      <input
        className="css-cg9o93-Checkbox"
        id="taskStatus"
        name="done"
        onChange={[Function]}
        placeholder=" "
        type="checkbox"
        value={false}
      />
      <label
        className="css-174x1gu-Checkbox"
        htmlFor="taskStatus"
      >
        in progress
      </label>
    </div>
    <div
      className="input-slider-wrapper css-1lncui6-InputSlider"
      data-testid="input-slider"
    >
      <input
        className="css-1wvyecb-InputSlider"
        id="taskTitle"
        name="title"
        onChange={[Function]}
        placeholder=" "
        required={true}
        type="text"
        value=""
      />
      <label
        className="css-f515j6-InputSlider"
        htmlFor="taskTitle"
      >
        task title
      </label>
    </div>
    <div
      className="input-slider-wrapper css-1lncui6-InputSlider"
      data-testid="input-slider"
    >
      <input
        className="css-1wvyecb-InputSlider"
        id="taskDescription"
        name="description"
        onChange={[Function]}
        placeholder=" "
        type="text"
        value=""
      />
      <label
        className="css-f515j6-InputSlider"
        htmlFor="taskDescription"
      >
        task description
      </label>
    </div>
    <button
      className="css-ctqbc-FormButton"
    >
      Add task
    </button>
     
  </form>
</div>
`;
